---
title: "A large-scale genome-wide association study of early-onset Alzheimer disease"
author: "Achal Neupane"
date: 2020-12-01T21:13:14-05:00
categories: ["Shell"]
tags: ["GWAS", "Alzheimer"]
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE)
```

# ADGC_2021_Analysis

## Background
Early-onset (≤65 years old) Alzheimer Disease (EOAD), which accounts for approximately 10% of all AD cases, is sometimes equated with dominantly inherited forms of AD (i.e., caused by known mutations in APP, PSEN1 or PSEN2). However, 90% of EOAD cases remain unexplained by these variants. This project aims to identify novel EOAD-associated variants through a large-scale, multi-ethnic genome-wide association (GWA) study.
Methods (90 words)
We have leveraged the Alzheimer Disease Genetics Consortium (ADGC) GWAS dataset that is aligned to GRCh38 and imputed with the TOPMed imputation server. This includes 39 studies from non-Hispanic Whites (NHW; CA=3,202; CO=5,782), 17 from African American (CA=139; CO=253), six from Asian (CA=133; CO=474), and seven from Hispanic (CA=199; CO=59) cohorts. Only the younger (≤65yo) and older participants (>80yo) were included for analysis as cases and controls, respectively. The genetic background was confirmed using PCA. We are performing single-variant analysis to identify variants that confer a higher risk for EOAD. 
Result (103 words)
Our preliminary single-variant association analysis (using sex, PC1-PC10 as covariates) emerging from the largest NHW cohort (CA=3,202; CO=5,782) have revealed nine different loci that were associated with EOAD at a suggestive P-value<1×10-6. The strongest signal from our analysis was in the APOE region. Additionally, we identified three novel genome-wide significant (P-value <5×10-8) loci of interest at HLA-DRB9 (rs9268852, P-value=2.9×10-9, OR=1.22); FAM86B3P (rs73199790, P-value=3x10-11, OR=1.32); and MSMB (rs2075895, P-value=2.83x10-16, OR=1.26); while also confirming the previously identified loci at MS4A4A (rs7108663, P-value=1.13x10-10, OR=0.80) and CR1 (rs6656401, P-value=2.54x10-12, OR=1.32). We will perform qualitative associations on the other three ethnicities to confirm the transferability of these signals.
Conclusion (96 words)
Our single-variant analysis has identified three novel loci associated with EOAD. This study is the largest genetic screening for risk and protective variants in EOAD and will be instrumental to identifying novel variants and pathways implicated in unexplained EOAD. Our future analysis includes gene-based association and pathway analyses per ethnicity to identify variants that confer a higher risk for EOAD, followed by transethnic analysis to validate the comprehensive nature of these risk factors. Additionally, we have also generated GWAS data from the Knight-ADRC participants, which will be merged with the ADGC datasets for our future analysis.




Here, I am performing PCA analysis with all four ethnicities.

WORKING_DIR =/40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/01-EOAD-preQC/02-Analysis/ADGC-HapMap-PCA/ADGC

Create symlinks to the plink files:

```r
# NHW
ln -s /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_NHW/ADGC_NHW_Cohort.fam ADGC_NHW_Cohort_without_chr.fam
ln -s /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_NHW/ADGC_NHW_Cohort.bed ADGC_NHW_Cohort_without_chr.bed
sed 's/chr//g' /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_NHW/ADGC_NHW_Cohort.bim >  ADGC_NHW_Cohort_without_chr.bim

# AA
ln -s /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_AA/ADGC_AA_Cohort.fam ADGC_AA_Cohort_without_chr.fam
ln -s /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_AA/ADGC_AA_Cohort.bed ADGC_AA_Cohort_without_chr.bed
sed 's/chr//g' /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_AA/ADGC_AA_Cohort.bim > ADGC_AA_Cohort_without_chr.bim

# Hispanic
ln -s /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_Hispanic/ADGC_Hispanic_Cohort.fam ADGC_Hispanic_Cohort_without_chr.fam
ln -s /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_Hispanic/ADGC_Hispanic_Cohort.bed ADGC_Hispanic_Cohort_without_chr.bed
sed 's/chr//g' /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_Hispanic/ADGC_Hispanic_Cohort.bim > ADGC_Hispanic_Cohort_without_chr.bim

# Asian
ln -s /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_Asian/ADGC_Asian_Cohort.fam ADGC_Asian_Cohort_without_chr.fam 
ln -s /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_Asian/ADGC_Asian_Cohort.bed ADGC_Asian_Cohort_without_chr.bed
sed 's/chr//g' /40/AD/GWAS_data/Source_Plink/2021_ADGC_EOAD/ADGC_Asian/ADGC_Asian_Cohort.bim > ADGC_Asian_Cohort_without_chr.bim
```

[HapMap PCA](ADGC_2021_Analysis%20cdfc76305a2746a1a8def50f783e624b/HapMap%20PCA%20aa8082bcb6534bd9bbf1614ea988d1fb.md)

[PCA 1000G](ADGC_2021_Analysis%20cdfc76305a2746a1a8def50f783e624b/PCA%201000G%20a964bd5de0a94d4fac05a278c81ea05d.md)

[Selection of samples by Age criteria](ADGC_2021_Analysis%20cdfc76305a2746a1a8def50f783e624b/Selection%20of%20samples%20by%20Age%20criteria%20a366d3e5e3dc4131bbdfe9f37a6d643c.md)

[2022-01-14 Single variant analyses - NHW](ADGC_2021_Analysis%20cdfc76305a2746a1a8def50f783e624b/2022-01-14%20Single%20variant%20analyses%20-%20NHW%20a98cd4b0c5904edf9d326292e75cd59b.md)
